tuist/tuist
# Add enforceExplicitDependendencies generation option

### Short description 📝

#### The description of the problem

I prepared a fixture `ios_app_with_implicit_dependencies` that has the following graph:
<img width="992" alt="image" src="https://github.com/tuist/tuist/assets/9371695/9477598f-f832-43ed-af04-62a3016366d9">

What we're trying to enforce is that the relation between `FrameworkA` and `FrameworkB` is explicit. This is currently not the case due to all products being built into the same directory. Let's say we build `FrameworkB` first. The `Build/Products` directory will look like this:
- `Build/Products`
  - `FrameworkB.framework`

Afterwards, when we build `FrameworkA.framework`, Xcode uses the combination of `BUILT_PRODUCTS_DIR` and `FRAMEWORK_SEARCH_PATHS` to find a particular framework.

Currently, `FRAMEWORK_SEARCH_PATHS` is empty _but_ `BUILT_PRODUCTS_DIR` points directly to `Build/Products`. Since we built `FrameworkB` separately before building `FrameworkA`, `FrameworkA` successfully finds the `FrameworkB.framework` artifact in `Build/Products` and the build succeeds – even when there's no _explicit_ relation between `FrameworkA` and `FrameworkB`.

#### Proposed solution

Soo, how do we propose to enforce describing the relation between `FrameworkA` and `FrameworkB` explicitly? By abandoning the Xcode's default behavior of dumping all built products into a single directory in `Build/Products`. Instead, we build frameworks into scoped directories.

To define _where_ built artifacts are saved into, we can use the [TARGET_BUILD_DIR](https://developer.apple.com/documentation/xcode/build-settings-reference#TARGETBUILDDIR) setting. The default value for this setting is `$(CONFIGURATION_BUILD_DIR)$(TARGET_BUILD_SUBPATH)`. To scope the artifacts, we will add an additional subpath equal to the product name: `$(CONFIGURATION_BUILD_DIR)$(TARGET_BUILD_SUBPATH)/$(PRODUCT_NAME)`.

This is how the `Build/Products` directory looks like after building `FrameworkB` now:
- `Build/Products`
  - `FrameworkB`
    - `FrameworkB.framework`

As described previously, Xcode tries to find available frameworks and build artifacts with the combination of `BUILT_PRODUCTS_DIR` and `FRAMEWORK_SEARCH_PATHS`. We also need to update these two settings, so that we can lookup only explicitly defined dependencies.

`BUILT_PRODUCTS_DIR` is updated to the same value is `TARGET_BUILD_DIR`: `$(CONFIGURATION_BUILD_DIR)$(TARGET_BUILD_SUBPATH)/$(PRODUCT_NAME)`. That way, we ignore the root `Build/Products` for lookup of build artifacts. How will now `FrameworkC` successfully find `FrameworkB` – as it should, since the dependency is explicit?

For each explicit dependency, we add an entry in `FRAMEWORK_SEARCH_PATHS`, for `FrameworkC` that means `$(CONFIGURATION_BUILD_DIR)$(TARGET_BUILD_SUBPATH)/FrameworkB`.

To recap, these will be the build settings for individual modules of the graph:
- `FrameworkB`
  - `TARGET_BUILD_DIR`: `Build/Products/FrameworkB`
  - `BUILT_PRODUCTS_DIR`:`Build/Products/FrameworkB`
  - `FRAMEWORK_SEARCH_PATHS`: Empty
- `FrameworkA`
  - `TARGET_BUILD_DIR`: `Build/Products/FrameworkA`
  - `BUILT_PRODUCTS_DIR`:`Build/Products/FrameworkA`
  - `FRAMEWORK_SEARCH_PATHS`: Empty (`FrameworkB` _won't be found_)
- `FrameworkC`
  - `TARGET_BUILD_DIR`: `Build/Products/FrameworkC`
  - `BUILT_PRODUCTS_DIR`:`Build/Products/FrameworkC`
  - `FRAMEWORK_SEARCH_PATHS`:`Build/Products/FrameworkB` (`FrameworkB` will be found)

#### Ensuring `.app` and other end products still work

We do these adjustments only for a couple of products:
- `dynamicLibrary`, `staticLibrary`, `framework`, `staticFramework`, `bundle`

I _did_ try doing the same adjustments for all products but this turned out be slightly more involved. For example, `App` can't rely on `FRAMEWORK_SEARCH_PATHS` only, but it actually needs all the built products (`.framework`, etc.) in the single directory specified by `BUILT_PRODUCTS_DIR`. We would need to copy or symlink all the individual directories of its transitive dependencies. While possible, this has the following drawbacks:
- The `Copy built products` could have overlapping input and output paths if you had a scheme with two apps, leading to errors
- Just copying those products is not enough for some scenarios – such as when running the host app

I fully expect most missing links to be between individual frameworks, not the final App target. That's why I decided for the following solution for end products like `.app`:

Each module with a scoped `TARGET_BUILD_DIR` also symlinks its built products into the default `Build/Products` directory. `App` keeps using this default directory, so it will find all necessary products – but other frameworks will still use only its scoped directory with framework search paths.

This does mean each framework will have an extra post build script – but we have an explicit set of input and output paths and we only override the symlink when the symlink doesn't exist. That should ensure the incremental compilation is not impacted. Additionally, these copied built products are only used by end products – other frameworks depend on `FRAMEWORK_SEARCH_PATHS` instead. 

#### External frameworks

While _ideally_ the graph for external frameworks would be valid, there's not much developers can do if explicit links are missing. I believe it's better to scope all external frameworks into a single `Build/Products/External` directory that internal targets use for lookup, so the explicitness is not required for them.

### How to test the changes locally 🧐

Run `tuist generate` on one of the fixtures or your own project and ensure that it works as expected (you can build it, run it, etc.)

### Contributor checklist ✅


### Reviewer checklist ✅
